### PR TITLE
Refactor NodeStack APIs to better handle the increase of distinct ID types.

### PR DESCRIPTION
**Pull request by jonmeow**
I'm looking at adding CallableId, so figured I'd do this API refactoring which should reduce code duplication. This increases the amount of cross-calls between APIs because it may not be an issue for performance after optimizations, and should simplify reading of the API.

Also note the prior static_asserts on layout were missing a couple types, which is why I'm moving them into Entry where it's going to be more obvious when something's added.

### CODE REVIEW

File: toolchain/semantics/semantics_handle.cpp [hunk 1]
```diff
@@ -70,10 +71,10 @@ auto SemanticsHandleDesignatedName(SemanticsContext& context,
 
 auto SemanticsHandleDesignatorExpression(SemanticsContext& context,
                                          ParseTree::Node parse_node) -> bool {
-  auto [_, name_id] = context.node_stack().PopForParseNodeAndNameId(
+  auto [_, name_id] = context.node_stack().PopWithParseNode<SemanticsStringId>(
       ParseNodeKind::DesignatedName);
```
Comment (zygoloid) [inline toolchain/semantics/semantics_handle.cpp]: It seems strange to pop with parse node then discard the parse node. Can you:
```suggestion
  auto name_id = context.node_stack().Pop<SemanticsStringId>(
      ParseNodeKind::DesignatedName);
```

File: toolchain/semantics/semantics_handle_struct.cpp [hunk 2]
```diff
@@ -47,8 +49,8 @@ auto SemanticsHandleStructFieldUnknown(SemanticsContext& context,
 auto SemanticsHandleStructFieldValue(SemanticsContext& context,
                                      ParseTree::Node parse_node) -> bool {
   auto [value_parse_node, value_node_id] =
-      context.node_stack().PopForParseNodeAndNodeId();
-  auto [_, name_id] = context.node_stack().PopForParseNodeAndNameId(
+      context.node_stack().PopWithParseNode<SemanticsNodeId>();
+  auto [_, name_id] = context.node_stack().PopWithParseNode<SemanticsStringId>(
       ParseNodeKind::DesignatedName);
```
Comment (zygoloid) [inline toolchain/semantics/semantics_handle_struct.cpp]: Same here:
```suggestion
  auto name_id = context.node_stack().Pop<SemanticsStringId>(
      ParseNodeKind::DesignatedName);
```

**Review by zygoloid — APPROVED** [review]

File: toolchain/semantics/semantics_handle.cpp (hunk 1 above)
Comment (jonmeow) [reply toolchain/semantics/semantics_handle.cpp]: Done

File: toolchain/semantics/semantics_handle_struct.cpp (hunk 2 above)
Comment (jonmeow) [reply toolchain/semantics/semantics_handle_struct.cpp]: Done

**Pull request merged**